src/io/unix/iouring.rs

```diff
@@ -0,0 +1,198 @@
+    }
+
+    /// Create new async iodriver from existing rio instance
+    pub fn new_async_rio(rio: Rio) -> Self {
+        Self {
+            rio: Some(rio),
+            sync_driver: SyncDriver::new_sync(),
+        }
+    }
+
+    /// Create new iodriver with async if available    
+    pub fn new_async_fallback_sync() -> Self {
+        if let Ok(rio) = rio::new() {
+            Self::new_async_rio(rio)
+        } else {
+            Self::new_sync()
+        }
+    }
+
+    pub(crate) async fn open(&self, path: impl AsRef<Path>) -> IOResult<File> {
+        let sync = self.sync_driver.open(path).await?;
+        File::from_file(sync, self.rio.clone()).await
+    }
+
+    pub(crate) async fn create(&self, path: impl AsRef<Path>) -> IOResult<File> {
+        let sync = self.sync_driver.create(path).await?;
+        File::from_file(sync, self.rio.clone()).await
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct File {
+    sync: SyncFile,
+    rio: Option<Rio>,
+}
+
+impl File {
+    pub(crate) fn size(&self) -> u64 {
+        self.sync.size()
+    }
+
+    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
+        if let Some(ref rio) = self.rio {
+            let offset = self.sync.file_size_append(buf.len() as u64);
+            self.write_all_at_aio(offset, buf, rio).await
+        } else {
+            self.sync.write_append_all(buf).await
+        }
+    }
+
+    pub(crate) async fn write_append_all_buffers(
+        &self,
+        first_buf: Bytes,
+        second_buf: Bytes,
+    ) -> IOResult<()> {
+        if let Some(ref rio) = self.rio {
+            let first_len = first_buf.len() as u64;
+            let second_len = second_buf.len() as u64;
+            let mut offset = self.sync.file_size_append(first_len + second_len);
+            self.write_all_at_aio(offset, first_buf, rio).await?;
+            offset = offset + first_len;
+            self.write_all_at_aio(offset, second_buf, rio).await
+        } else {
+            self.sync
+                .write_append_all_buffers(first_buf, second_buf)
+                .await
+        }
+    }
+
+    pub(crate) async fn write_all_at(&self, offset: u64, buf: Bytes) -> IOResult<()> {
+        debug_assert!(offset + buf.len() as u64 <= self.size());
+        if let Some(ref rio) = self.rio {
+            self.write_all_at_aio(offset, buf, rio).await
+        } else {
+            self.sync.write_all_at(offset, buf).await
+        }
+    }
+
+    pub(crate) async fn read_all(&self) -> Result<BytesMut> {
+        self.read_exact_at_allocate(self.size().try_into()?, 0)
+            .await
+    }
+
+    pub(crate) async fn read_exact_at_allocate(
+        &self,
+        size: usize,
+        offset: u64,
+    ) -> Result<BytesMut> {
+        let buf = BytesMut::zeroed(size);
+        self.read_exact_at(buf, offset).await
+    }
+
+    pub(crate) async fn read_exact_at(&self, mut buf: BytesMut, offset: u64) -> Result<BytesMut> {
```

Comment:
There is a warning about `mut`

```suggestion
    pub(crate) async fn read_exact_at(&self, buf: BytesMut, offset: u64) -> Result<BytesMut> {
```